### PR TITLE
Update the Crunchy-Postgres statefulset specs to successfully deploy replication cluster

### LINKED ISSUE #461: Provide steps to test application HA on crunchy postgres containers on k8s w/ OpenEBS storage
## Is this a BUG REPORT or FEATURE REQUEST?
 FEATURE REQUEST

Currently, we have the statefulset-specification JSON files to deploy the postgreSQL cluster. We need to identify steps to generate load from a postgres client and demonstrate the application HA, with the client job continuing despite loss of access to pod.

### PR DESCRIPTION
**Pull request by ksatchit**
**What this PR does / why we need it**:

 The current JSON specs for Crunchy-Postgres statefulset fail to bring up the database cluster. Following changes have been made to enable this: 

- Changed the PG_MODE env variable in the spec from "Master" to "Set". Latter is needed by the startup scripts to decide on instance role based on ordinality

- Included Security Context for the pods to allow a user generated on the container to access and manage the mounted volume

- Included PG_PRIMARY_PORT env variable for successful port registration

- Included PG_PRIMARY_HOST env variable. This is needed by the slave/replica to sync with the primary/master

- Updated Crunchy-Postgres image version to latest stable release 1.6.0

- Updated semantics changes as needed by 1.6.0 (master->primary)

- Updated README with deployment, usage and verification instructions

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #461 

**Special notes for your reviewer**:

Refer https://github.com/CrunchyData/crunchy-containers/issues/287

### CODE REVIEW

**Comment by ksatchit** [comment]
Trailing log snippet on the pgset-0 & pgset-1 database pods
-------------------------------------------------------------------
pgset-0/primary: 

```
2017-11-14 06:36:20 EST [85]: [1-1] user=,db=,app=,client=LOG:  listening on IPv4 address "0.0.0.0", port 5432
2017-11-14 06:36:20 EST [85]: [2-1] user=,db=,app=,client=LOG:  listening on IPv6 address "::", port 5432
2017-11-14 06:36:20 EST [85]: [3-1] user=,db=,app=,client=LOG:  listening on Unix socket "/tmp/.s.PGSQL.5432"
2017-11-14 06:36:20 EST [85]: [4-1] user=,db=,app=,client=LOG:  redirecting log output to logging collector process
2017-11-14 06:36:20 EST [85]: [5-1] user=,db=,app=,client=HINT:  Future log output will appear in directory "pg_log".
```

pgset-1/replica: 

```
Attempting pg_isready on primary
+ true
+ psql -h pgset-primary -p 5432 -U primaryuser userdb -f /opt/cpm/bin/readiness.sql
psql: FATAL:  the database system is shutting down
+ '[' 2 -eq 0 ']'
+ echo 'Attempting pg_isready on primary '
Attempting pg_isready on primary
+ sleep 2
+ true
+ psql -h pgset-primary -p 5432 -U primaryuser userdb -f /opt/cpm/bin/readiness.sql
              now
-------------------------------
 2017-11-14 06:36:21.233572-05
(1 row)

+ '[' 0 -eq 0 ']'
+ echo 'The database is ready.'
The database is ready.
+ CONNECTED=true
+ break
+ pg_basebackup -X fetch --no-password --pgdata /pgdata/pgset-1 --host=pgset-primary --port=5432 -U primaryuser
+ [[ -v SYNC_REPLICA ]]
+ APPLICATION_NAME=pgset-1
+ echo 'SYNC_REPLICA not set'
SYNC_REPLICA not set
+ echo pgset-1 ' is the APPLICATION_NAME being used'
pgset-1  is the APPLICATION_NAME being used
+ cp /opt/cpm/conf/pgrepl-recovery.conf /tmp
+ sed -i s/PG_PRIMARY_USER/primaryuser/g /tmp/pgrepl-recovery.conf
+ sed -i s/PG_PRIMARY_HOST/pgset-primary/g /tmp/pgrepl-recovery.conf
+ sed -i s/PG_PRIMARY_PORT/5432/g /tmp/pgrepl-recovery.conf
+ sed -i s/APPLICATION_NAME/pgset-1/g /tmp/pgrepl-recovery.conf
+ cp /tmp/pgrepl-recovery.conf /pgdata/pgset-1/recovery.conf
+ '[' -f /pgconf/postgresql.conf ']'
+ date
Tue Nov 14 11:36:31 UTC 2017
+ [[ -v PGAUDIT_ANALYZE ]]
+ wait
+ postgres -D /pgdata/pgset-1
2017-11-14 06:36:31 EST [45]: [1-1] user=,db=,app=,client=LOG:  listening on IPv4 address "0.0.0.0", port 5432
2017-11-14 06:36:31 EST [45]: [2-1] user=,db=,app=,client=LOG:  listening on IPv6 address "::", port 5432
2017-11-14 06:36:31 EST [45]: [3-1] user=,db=,app=,client=LOG:  listening on Unix socket "/tmp/.s.PGSQL.5432"
2017-11-14 06:36:31 EST [45]: [4-1] user=,db=,app=,client=LOG:  redirecting log output to logging collector process
2017-11-14 06:36:31 EST [45]: [5-1] user=,db=,app=,client=HINT:  Future log output will appear in directory "pg_log".
```

File: k8s/demo/crunchy-postgres/README.md [hunk 1]
```diff
@@ -1,35 +1,203 @@
-This is place-holder for providing the steps to run clustered postgresql on K8s+OpenEBS. We hope to merge these files back to https://github.com/CrunchyData/crunchy-containers/blob/master/examples/kube/
+# Running Crunchy-Postgres with OpenEBS
 
+This tutorial provides detailed instructions to run a PostgreSQL Statefulset with OpenEBS storage and perform
```
Comment (mahebbar) [inline k8s/demo/crunchy-postgres/README.md]: Statefulset = StatefulSet

File: k8s/demo/crunchy-postgres/README.md [hunk 2]
```diff
@@ -1,35 +1,203 @@
-This is place-holder for providing the steps to run clustered postgresql on K8s+OpenEBS. We hope to merge these files back to https://github.com/CrunchyData/crunchy-containers/blob/master/examples/kube/
+# Running Crunchy-Postgres with OpenEBS
 
+This tutorial provides detailed instructions to run a PostgreSQL Statefulset with OpenEBS storage and perform 
+simple database operations to verify successful deployment. 
 
+## Crunchy-Postgres 
 
-## Usage
+The postgres container used in the Statefulset is sourced from (CrunchyData)[https://github.com/CrunchyData/crunchy-containers]. CrunchyData provides cloud agnostic PostgreSQL container technology that is designed for production
```
Comment (mahebbar) [inline k8s/demo/crunchy-postgres/README.md]: Statefulset = StatefulSet (all occurrences to be changed)

File: k8s/demo/crunchy-postgres/README.md [hunk 3]
```diff
@@ -1,35 +1,203 @@
...
+# Running Crunchy-Postgres with OpenEBS
 
+This tutorial provides detailed instructions to run a PostgreSQL Statefulset with OpenEBS storage and perform 
+simple database operations to verify successful deployment. 
 
+## Crunchy-Postgres 
 
-## Usage
+The postgres container used in the Statefulset is sourced from (CrunchyData)[https://github.com/CrunchyData/crunchy-containers]. CrunchyData provides cloud agnostic PostgreSQL container technology that is designed for production 
+workloads with cloud native HA, DR, and monitoring.
```
Comment (mahebbar) [inline k8s/demo/crunchy-postgres/README.md]: DR = expand?

File: k8s/demo/crunchy-postgres/README.md [hunk 4]
```diff
@@ -1,35 +1,203 @@
...
+storage classes.
+
+```
+test@Master:~/crunchy-postgres$ kubectl get pods
+NAME                                                             READY     STATUS    RESTARTS   AGE
+maya-apiserver-2245240594-ktfs2                                  1/1       Running   0          3h
+openebs-provisioner-4230626287-t8pn9                             1/1       Running   0          3h
+```
+
+## Deploy the Crunchy-Postgres Statefulset with OpenEBS storage
```
Comment (mahebbar) [inline k8s/demo/crunchy-postgres/README.md]: storage = Storage (Title case for all headings)

File: k8s/demo/crunchy-postgres/README.md [hunk 5]
```diff
@@ -1,35 +1,203 @@
...
+
+The Statefulset specification JSONs are available at OpenEBS/k8s/demo/crunchy-postgres. 
+
+The number of replicas in the Statefulset can be modified in the *set.json* file. This example uses 2 replicas, 
+which includes one master and one slave. The Postgres pods are configured as primary/master or as replica/slave by 
+a startup script which decides the role based on ordinality assigned to the pod.
```
Comment (mahebbar) [inline k8s/demo/crunchy-postgres/README.md]: ordinality ? Need to clarify with you. Will discuss.

File: k8s/demo/crunchy-postgres/README.md [hunk 6]
```diff
@@ -1,35 +1,203 @@
...
+service "pgset" created
++ kubectl create -f /home/test/openebs/k8s/demo/crunchy-postgres/set-primary-service.json
+service "pgset-primary" created
++ kubectl create -f /home/test/openebs/k8s/demo/crunchy-postgres/set-replica-service.json
+service "pgset-replica" created
++ kubectl create -f /home/test/openebs/k8s/demo/crunchy-postgres/set.json
+statefulset "pgset" created
+```
+
+Verify that all the OpenEBS persistent volumes are created, the Crunchy-Postgres services and pods are running.
```
Comment (mahebbar) [inline k8s/demo/crunchy-postgres/README.md]: are created and the Crunchy-Postgres services and pods are running.

File: k8s/demo/crunchy-postgres/README.md [hunk 7]
```diff
@@ -1,35 +1,203 @@
...
+
+Note: It may take some time for the pods to start as the images must be pulled and instantiated. This is also
+dependent on the network speed.
+
+
+## Verify Successful Crunchy-Postgres Deployment
+
+The verification procedure can be carried out using the following steps: 
+
+- Check cluster replication status between the Postgres primary and replica
```
Comment (mahebbar) [inline k8s/demo/crunchy-postgres/README.md]: Postgres primary and replica clusters.

File: k8s/demo/crunchy-postgres/README.md [hunk 8]
```diff
@@ -1,35 +1,203 @@
...
+- Create a table in the default database as a Postgres user on the primary
```
Comment (mahebbar) [inline k8s/demo/crunchy-postgres/README.md]: As a Postgres user, create a table in the default database on the primary cluster.

File: k8s/demo/crunchy-postgres/README.md [hunk 9]
```diff
@@ -1,35 +1,203 @@
...
+- Check data synchronization on the replica for created table
```
Comment (mahebbar) [inline k8s/demo/crunchy-postgres/README.md]: Check data synchronization on the replica cluster in the table you have created.

File: k8s/demo/crunchy-postgres/README.md [hunk 10]
```diff
@@ -1,35 +1,203 @@
...
+- Verify table creation is restricted on the replica
```
Comment (mahebbar) [inline k8s/demo/crunchy-postgres/README.md]: Verify that the table is not created on the replica cluster.

File: k8s/demo/crunchy-postgres/README.md [hunk 11]
```diff
@@ -1,35 +1,203 @@
...
+(1 row)
 ```
 
-### Step 2: Run the StatefulSet. 
-The files are make available with default images and credentails (set.json)
+The replica should be registered for "asynchronous" replication. 
+
+### Step-3: Create a Table with Test Content on the Default Database 
+
+These queries should be executed on the primary.
```
Comment (mahebbar) [inline k8s/demo/crunchy-postgres/README.md]: primary cluster.

File: k8s/demo/crunchy-postgres/README.md [hunk 12]
```diff
@@ -1,35 +1,203 @@
...
 ```
+```
+test@Master:~/crunchy-postgres$ psql -h 10.47.0.3 -U testuser postgres -c 'insert into foo values (1)'
+Password for user testuser:
+INSERT 0 1
+```
+
+### Step-4: Verify Data Synchronization on Replica
+
+Identify the IP address of the replica (pgset-1) pod or the service (pgset-replica) and execute the following
```
Comment (mahebbar) [inline k8s/demo/crunchy-postgres/README.md]: IP Address
following command.

File: k8s/demo/crunchy-postgres/README.md [hunk 13]
```diff
@@ -1,35 +1,203 @@
...
-kubectl get pv
+test@Master:~$ psql -h 10.44.0.6 -U testuser postgres -c 'table foo'
+Password for user testuser:
+ id
+----
+  1
+(1 row)
 ```
 
+Verify that the table content is replicated successfully
```
Comment (mahebbar) [inline k8s/demo/crunchy-postgres/README.md]: successfully.

File: k8s/demo/crunchy-postgres/README.md [hunk 14]
```diff
@@ -1,35 +1,203 @@
...
 
-## Motivation
+### Step-5: Verify Database Write Restricted on Replica
 
-The k8s spec files are based on the files provided by [CrunchyData StatefulSet with Dynamic Provisioner](https://github.com/CrunchyData/crunchy-containers/tree/master/examples/kube/statefulset-dyn)
+Attempt creation of a new table on the replica and verify the operation is unsuccessful.
```
Comment (mahebbar) [inline k8s/demo/crunchy-postgres/README.md]: Attempt to create a new table on the replica, and verify that the table creation is unsuccessful.

File: k8s/demo/crunchy-postgres/run.sh [hunk 15]
```diff
@@ -18,22 +18,22 @@
 
 DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"
 
-$DIR/cleanup.sh
+#$DIR/cleanup.sh
 
 # create the service account used in the containers
 kubectl create -f $DIR/set-sa.json
 
 # as of Kube 1.6, we need to allow the service account to perform
```
Comment (mahebbar) [inline k8s/demo/crunchy-postgres/run.sh]: For version Kube 1.6, we must allow the service account to perform a label command. For this example, OpenEBS opens up wide permissions for all service accounts. This is NOT for production!

File: k8s/demo/crunchy-postgres/README.md (hunk 5 above)
Comment (ksatchit) [reply k8s/demo/crunchy-postgres/README.md]: Ordinality is a property/value assigned by kubernetes to the constituent instances/pod of a statefulset. For example, in the pods below, 0 & 1 are the ordinality values. Based on the ordinality provided, the DB startup scripts configure it as primary/master OR as a slave/replica.

pgset-0
pgset-1

File: k8s/demo/crunchy-postgres/README.md (hunk 7 above)
Comment (ksatchit) [reply k8s/demo/crunchy-postgres/README.md]: Both primary and replica as members of a single cluster. So we address them as "primary" &  "replica" in the cluster context.

File: k8s/demo/crunchy-postgres/README.md (hunk 8 above)
Comment (ksatchit) [reply k8s/demo/crunchy-postgres/README.md]: Here, postgres user is a non-root linux "user" (in this case, testuser) created by the startup script in the postgres container. It is not a generic reference to a consumer/user of the postresql db. Modified this statement to clarify this.

**Comment by ksatchit** [comment]
@mahebbar , updated the files with review comments.

File: k8s/demo/crunchy-postgres/README.md [hunk 16]
```diff
@@ -1,35 +1,203 @@
...
+dependent on the network speed.
+
+
+## Verify Successful Crunchy-Postgres Deployment
+
+The verification procedure can be carried out using the following steps: 
+
+- Check cluster replication status between the Postgres primary and replica
+- Create a table in the default database as Postgres user "testuser" on the primary
+- Check data synchronization on the replica for table you have created
```
Comment (mahebbar) [inline k8s/demo/crunchy-postgres/README.md]: for table = for the table

File: k8s/demo/crunchy-postgres/README.md [hunk 17]
```diff
@@ -1,35 +1,203 @@
...
+- Verify that table is not created on the replica
+
+### Step-1: Install the PostgreSQL-Client
+
+Install the PostgreSQL CLient Utility (psql) on any of the Kubernetes machines to perform database operations 
+from command line.
```
Comment (mahebbar) [inline k8s/demo/crunchy-postgres/README.md]: from command line = from the command line.

File: k8s/demo/crunchy-postgres/README.md [hunk 18]
```diff
@@ -1,35 +1,203 @@
...
+Install the PostgreSQL CLient Utility (psql) on any of the Kubernetes machines to perform database operations 
+from command line. 
+
+```
+sudo apt-get install postgresql-client
+```
+
+### Step-2: Verify Cluster Replication Status on Crunchy-Postgres Cluster
+
+Identify the IP address of the primary (pgset-0) pod or the service (pgset-primary) and execute the following
```
Comment (mahebbar) [inline k8s/demo/crunchy-postgres/README.md]: IP address = IP Address

File: k8s/demo/crunchy-postgres/README.md [hunk 19]
```diff
@@ -1,35 +1,203 @@
...
+(1 row)
 ```
 
+Verify that the table content is replicated successfully.
 
-## Motivation
+### Step-5: Verify Database Write Restricted on Replica
 
-The k8s spec files are based on the files provided by [CrunchyData StatefulSet with Dynamic Provisioner](https://github.com/CrunchyData/crunchy-containers/tree/master/examples/kube/statefulset-dyn)
+Attempt to create a new table on the replica, and verify that the creation is unsuccessful.
```
Comment (mahebbar) [inline k8s/demo/crunchy-postgres/README.md]: Create a new table on the replica and verify that the table is not created.

File: k8s/demo/crunchy-postgres/run.sh [hunk 20]
```diff
@@ -18,22 +18,22 @@
...
-# as of Kube 1.6, we need to allow the service account to perform
-# a label command, for this example, we open up wide permissions
-# for all serviceaccounts, this is NOT for production!
-#kubectl create clusterrolebinding permissive-binding \
-#  --clusterrole=cluster-admin \
-#  --user=admin \
-#  --user=kubelet \
-#  --group=system:serviceaccounts
+# For version Kube 1.6, we must allow the service account to perform
+# a label command. For this example, OpenEBS open up wide permissions
```
Comment (mahebbar) [inline k8s/demo/crunchy-postgres/run.sh]: open = opens

File: k8s/demo/crunchy-postgres/run.sh [hunk 21]
```diff
@@ -18,22 +18,22 @@
...
+# for all service accounts, this is NOT for production!
```
Comment (mahebbar) [inline k8s/demo/crunchy-postgres/run.sh]: accounts. This is NOT...

File: k8s/demo/crunchy-postgres/README.md (hunk 19 above)
Comment (ksatchit) [reply k8s/demo/crunchy-postgres/README.md]: The expectation is that table should not be created. Which is why we are asking the user to *attempt* it. Felt the first part is an affirmative statement which we contradict immediately afterwards.

**Pull request merged**